sdcm/provision/aws/provisioner.py

```diff
@@ -0,0 +1,290 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2021 ScyllaDB
+
+import contextlib
+import datetime
+import logging
+import time
+from typing import List, Optional, Union
+
+from mypy_boto3_ec2 import EC2Client
+from mypy_boto3_ec2.service_resource import Instance
+
+from sdcm.provision.aws.instance_parameters import AWSInstanceParams
+from sdcm.provision.aws.utils import ec2_services, ec2_clients, find_instance_by_id, set_tags_on_instances, \
+    wait_for_provision_request_done, create_spot_fleet_instance_request, \
+    create_spot_instance_request, SPOT_CNT_LIMIT, SPOT_FLEET_LIMIT, SPOT_REQUEST_TIMEOUT, STATUS_FULFILLED, \
+    SPOT_STATUS_UNEXPECTED_ERROR, FLEET_LIMIT_EXCEEDED_ERROR, SPOT_CAPACITY_NOT_AVAILABLE_ERROR, MAX_SPOT_DURATION_TIME
+from sdcm.provision.common.provisioner import TagsType, ProvisionGuaranteedParams, ProvisionerBase, \
+    ProvisionNotGuaranteedParams, ProvisionParamsBase
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AWSProvisioner(ProvisionerBase):  # pylint: disable=too-few-public-methods
+    # TODO: Make them configurable
+    _wait_interval = 5
+    _iam_fleet_role = 'arn:aws:iam::797456418907:role/aws-ec2-spot-fleet-role'
+
+    def provision(  # pylint: disable=too-many-arguments
+            self,
+            provision_parameters: ProvisionParamsBase,
+            instance_parameters: AWSInstanceParams,
+            count: int,
+            tags: Union[List[TagsType], TagsType] = None,
+            names: List[str] = None) -> List[Instance]:
+        if tags is None:
+            tags = {}
+        if isinstance(tags, dict):
+            tags = [tags] * count
+        elif isinstance(tags, list):
+            tags = tags.copy()
+
+        if names:
+            assert len(names) == count, "Names length should be equal to count"
+        assert len(tags) == count, "Tags length should be equal to count"
+
+        for node_id, name in enumerate(names):
+            tag = tags[node_id]
+            tag['Name'] = name
+
+        if isinstance(provision_parameters, ProvisionGuaranteedParams):
```

Comment:
Why do you call it "Guaranteed" and not "on-demand"?

Reply:
`on-demand` is not quite telling name, while `guaranteed` means that it is guaranteed that you are going to get your instance.

Reply:
Well, I don’t won’t to pick on semantics, but even on-demand is not guaranteed.
Sometimes AWS cannot supply even on demand instances in some AZ.

I don’t mind we will decide on different terms than AWS to generalize it, but it actually looks like the other clouds are adopting AWS terms.

Reply:
let's rename in separate PR
ProvisionGuaranteedParams -> OnDemandParams
ProvisionNonGuaranteedParams - SpotParams